src/iocore/cache/CacheEvacuateDocVC.cc

```diff
@@ -57,88 +57,84 @@ CacheEvacuateDocVC::evacuateDocDone(int /* event ATS_UNUSED */, Event * /* e ATS
        (int)dir_phase(&this->dir));
   int i = dir_evac_bucket(&this->overwrite_dir);
   // nasty beeping race condition, need to have the EvacuationBlock here
-  EvacuationBlock *b = this->stripe->evac_bucket_valid(i) ? this->stripe->evacuate[i].head : nullptr;
-  for (; b; b = b->link.next) {
-    if (dir_offset(&b->dir) == dir_offset(&this->overwrite_dir)) {
-      // If the document is single fragment (although not tied to the vector),
-      // then we don't have to put the directory entry in the lookaside
-      // buffer. But, we have no way of finding out if the document is
-      // single fragment. doc->single_fragment() can be true for a multiple
-      // fragment document since total_len and doc->len could be equal at
-      // the time we write the fragment down. To be on the safe side, we
-      // only overwrite the entry in the directory if its not a head.
-      if (!dir_head(&this->overwrite_dir)) {
-        // find the earliest key
-        EvacuationKey *evac = &b->evac_frags;
-        for (; evac && !(evac->key == doc->key); evac = evac->link.next) {
-          ;
-        }
-        ink_assert(evac);
-        if (!evac) {
-          break;
-        }
-        if (evac->earliest_key.fold()) {
-          DDbg(dbg_ctl_cache_evac, "evacdocdone: evacuating key %X earliest %X", evac->key.slice32(0),
-               evac->earliest_key.slice32(0));
-          EvacuationBlock *eblock = nullptr;
-          Dir              dir_tmp;
-          dir_lookaside_probe(&evac->earliest_key, this->stripe, &dir_tmp, &eblock);
-          if (eblock) {
-            CacheEvacuateDocVC *earliest_evac  = eblock->earliest_evacuator;
-            earliest_evac->total_len          += doc->data_len();
-            if (earliest_evac->total_len == earliest_evac->doc_len) {
-              dir_lookaside_fixup(&evac->earliest_key, this->stripe);
-              free_CacheEvacuateDocVC(earliest_evac);
-            }
+  EvacuationBlock *b = this->stripe->evac_bucket_valid(i) ? this->stripe->get_preserved_dirs().find(this->overwrite_dir) : nullptr;
+  if (b) {
+    // If the document is single fragment (although not tied to the vector),
+    // then we don't have to put the directory entry in the lookaside
+    // buffer. But, we have no way of finding out if the document is
+    // single fragment. doc->single_fragment() can be true for a multiple
+    // fragment document since total_len and doc->len could be equal at
+    // the time we write the fragment down. To be on the safe side, we
+    // only overwrite the entry in the directory if its not a head.
+    if (!dir_head(&this->overwrite_dir)) {
+      // find the earliest key
+      EvacuationKey *evac = &b->evac_frags;
+      for (; evac && !(evac->key == doc->key); evac = evac->link.next) {
+        ;
+      }
+      ink_assert(evac);
+      if (!evac) {
+        goto RET;
```

Comment:
How about simply returning from here?

If calling `free_CacheEvacuateDocVC(this)` is the matter, setting defer by `ts::PostScript` in the beginning is another apporach.
```
ts::PostScript defer_free([&](){ free_CacheEvacuateDocVC(this); });
```

Reply:
In other branches we don't free it. Maybe the cleanest thing for now is to simply return as you say and also duplicate the call to free.